File: Dp/1531. [Like And Useful]String Compression II.py
// Very Smart solution...
// so just need to check last char.
// each have delete or not delete... some edge case check
// O(26*N^3)

class Solution:
    def getLengthOfOptimalCompression(self, s: str, k: int) -> int:
        @lru_cache(None)
        def helper(ind, last, last_cnt, left):
            if left < 0:
                return float('inf')
            if ind >= len(s):
                return 0
            if s[ind] == last:
                inc = 1 if last_cnt == 1 or last_cnt == 9 or last_cnt == 99 else 0
                return inc + helper(ind + 1, last, last_cnt+1, left)
            else:
                keep_char = 1 + helper(ind + 1, s[ind], 1, left) 
                del_char = helper(ind + 1, last, last_cnt, left - 1) 
                return min(keep_char, del_char)
        return helper(0, "", 0, k)

File: Dp/1531. [Like And Useful]String Compression II.cpp
// memo need to be global or memory limit.
// so just enum.
// the trick is the last char. and contan value.
// so each can delete or not delete

int dp[105][105][30][105];
class Solution {
public:
    string s;
    int dfs(int i, int k, int last, int con){
        if(k < 0) return 999999;
        if(i >= s.size()) return 0;
        if(dp[i][k][last][con] != -1) return dp[i][k][last][con];
        int res = 999999;
        res = min(res, dfs(i+1, k-1, last, con));
        if(s[i] - 'a' == last){
            int inc = (con == 1 || con == 9 || con == 99) ? 1 : 0;
            res = min(inc + dfs(i+1, k, last, con + 1), res);
        }
        else{
            res = min(1 + dfs(i+1, k, s[i]-'a', 1), res);
        }
        return dp[i][k][last][con] = res;
    }
    int getLengthOfOptimalCompression(string s, int k) {
        this->s = s;
        memset(dp, -1, sizeof dp);
        return dfs(0, k, 27, 0);
    }
};
